feat(KONFLUX-5434): ADR 47: Creating pending pipeline runs (#218)
konflux-ci/architecture: ADR/0039-creating-pending-pipelineruns.md

> # 39. Creating pending pipeline runs

Author: Thank you, fixed.

Date: 2024-11-21

## Status

Accepted

## Context

> There is a frequent use case amongst Konflux users when triggering large amount of builds: sometimes users do not care how quickly the pipeline run will finish, but they want it to finish eventually.
> Now what happens is that part of the pipeline runs started in this big bulk is blocked by unavailable resources (e.g. namespace memory limit quotas or MPC pool size).
> Because of how Tekton works, no matter what these pipeline runs are considered running (even if their pods are waiting for resources) and so eventually timeout setting strikes.

Reviewer: Trying to present the problem more clearly. Feel free to take any or none of the suggestion (some of this suggestion blurs into the next paragraph too).
```suggestion
When pipelines start, the time is immediately being taken away from the configured timeout (even if their pods are waiting for resources). While this behavior is likely sufficient when fewer concurrent pipelines are running, timeouts are more likely to be exceeded when bulk builds are triggered (i.e. a Renovate job updating common dependencies). While the speed at which a pipeline completes is important, a delayed and completed pipeline will likely finish quicker than an immediately and timeout-killed pipeline. Futhermore, having multiple pipelines running at the same time may starve other concurrent pipelines from acquiring the necessary resources resulting in fewer completed pipelines.
```

> So say you started 300 pipeline runs at Fry evening and you expect to see 300 new images at Mon morning.

Reviewer 2: ```suggestion
So say you started 300 pipeline runs on Friday evening and you expect to see 300 new images by Monday morning.
```

You have resource quota for 20 concurrent builds, so 20 builds get started, when they finish 20 more and so on.

> But because all 300 pipeline runs started (from Tekton point of view) when they were created, after few iterations all the remaining pipeline runs will be cancelled (as timeouted).

Reviewer 2: ```suggestion
But because all 300 pipeline runs started (from Tekton point of view) when they were created, after few iterations all the remaining pipeline runs will be cancelled (due to a timeout).
```

> But because all 300 pipeline runs started (from Tekton point of view) when they were created, after few iterations all the remaining pipeline runs will be cancelled (as timeouted).

Reviewer: As I mentioned above, the case could be that none of the pipelines completed because they were all starving each other of the needed resources.

Author: Hello! I have not experienced that, but it is definitely possible, especially in combination with some MPC bugs.

This is mine mental model of what is happening:

Say you have this pipeline:

1. init
2. clone
3. prefetch
4. [build-s390x, build-arm64]

The pipeline usually runs for 1 hour and have timeout set to 2 hours. MPC have capacity of 2 s390x and 2 arm64, all free.

Now you trigger 5 builds.

1. First 2 builds start, all 5 builds are considered started by Tekton, last 3 are waiting for VMs
2. Hour later, first 2 builds finish, second 2 builds start and last one is still waiting
3. Another hour later second 2 builds finish (just in time before timeout) and last, 5th, build fails because of timeout

So we end up with 4 finished builds and one failed because of timeout.

Anyway, I'm happy to change the wording as I think we both agree that the issue is there.

> Another use case for what we are proposing is protecting cluster (e.g. etcd) from spikes of running lots of concurrent pipeline runs.

Reviewer: ```suggestion
When many pipelines are triggered across multiple namespaces, there is also an increased load on etcd which may degrade the cluster's overall performance.
```

## Decision

> We will use idea of creating pipeline runs in pending state and having a standalone controller that would implement any custom logic on when to start each pending pipeline:

Reviewer: Is this proposed as a Konflux core or addon component?

Many of the considerations refer to the multi-platform controller, but that is not a core component.

Author: Oh, I think it depends on what logic we will use. If we decide we want to protect etcd by capping number of concurrent pipelineruns, it can be core.

If we decide we want to protect multi-arch users from timeouting, it should be addon I guess.

Should I add this to the doc?

> 1. We add `.spec.status: "PipelineRunPending"` to pipeline run YAMLs.

Reviewer: The numbering is off in this section.

Author: Should be OK now.

1. We will implement controller, that will contain code that will decide when to start (by just nulling `.spec.status` field - then Pipelines Controller takes over) what pipeline run. Some possibilities come to mind:

>    1. Controller will keep an eye on current number of concurrently running pipeline runs in the cluster (using Prometheus metric `avg(tekton_pipelines_controller_running_pipelineruns)`), and if that is below given threshold, it starts the pipeline run(s).

Reviewer: How might this be reconciled between namespace quotas and cluster-wide performance?

Author: I would expect that a pending pipeline run need to pass all the gates (AND) to be actually started. Say that MPC and etcd-protection gate applies to it: it need to pass both of them to be started.

But I'm unsure if this answers your question, sorry :)

>    1. If pipeline run have `prsstarter.example.com/mpc_platforms: linux/arm64,linux/ppc64le,linux/s390x` annotation, controller checks if MPC have free capacity (using max for the pool and Prometheus metric `multi_platform_controller_running_tasks`) for mentioned platforms and if so, will start it.

Reviewer: Is this an annotation that is used today or is it a new proposed one? Can it be generalized as the multi-platform-controller isn't a core component?

The specific annotation values are implementations of individual Konflux deployments.

Author: It is just a random string I used in tests in Stage we did.

I was thinking about this as sort of plugable system:

* `prsstarter.example.com/mpc_platforms: linux/arm64,linux/ppc64le,linux/s390x` - means you want your pipeline run to be gated based on MPC pool availability
* `prsstarter.example.com/memory_limit_quota: 10GiB` - means you want your pipeline to be started only if your namespace have at least 10GiB free in memory limit quota
* `prsstarter.example.com/other_plugin: other_plugin_configuration` - means you want your pipeline to be gated by some other plugin

Should I just avoid specific annotations or stress it is just an example? How can I improve it in the doc?

2024-11-20 14:04:59,660 prs_starter.my_reconciler Thread-2 (my_reconciler) INFO Starting my-comp-2-on-push-zz6sw/jhutar-tenant (429faf3a-620d-4ccd-ac89-c7ba4db11c90)
2024-11-20 14:05:00,529 prs_starter.my_reconciler Thread-2 (my_reconciler) INFO The my-comp-2-on-push-zz6sw/jhutar-tenant (429faf3a-620d-4ccd-ac89-c7ba4db11c90) was started, not watching it anymore
2024-11-20 14:05:00,529 prs_starter.my_reconciler Thread-2 (my_reconciler) DEBUG Reconciling my-comp-3-on-push-hj5bw/jhutar-tenant (ba328d32-e912-4a2c-98e8-506937577912)
2024-11-20 14:05:00,529 prs_starter.my_reconciler Thread-2 (my_reconciler) DEBUG Missing capacity for my-comp-3-on-push-hj5bw/jhutar-tenant (ba328d32-e912-4a2c-98e8-506937577912) on linux/s390x
```

This is what was happening:

* at 2024-11-20 14:04:44,646, we checked MPC capacity and s390x was fully used, so reconciling pipeline runs (that are in pending state) `my-comp-2-on-push-...` and `my-comp-3-on-push-...` did not started them
* at 2024-11-20 14:04:57,309, we noticed MPC freed the capacity for s390x

> * at 2024-11-20 14:04:59,660 pipeline run `my-comp-2-on-push-...` was started, no luck for `my-comp-3-on-push-...` though

Reviewer: These types of capacity decisions are not always blocked on single events/quotas. When considering the multi-platform controller, there are multiple different platforms that may be requested. 

* How can we reconcile the queues for all of these platforms?
* It seems like we would need to add some sort of queuing mechanism for the pending PipelineRuns. How can we ensure fairness with all of the pipelines which may be waiting for different metrics/resources to indicate capacity?
* How can we ensure that a pipeline is not unnecessarily blocked by an unneeded resource just because it is waiting behind some other pipeline in a queue?

Author: > How can we reconcile the queues for all of these platforms?

Here I was checking all platforms from that `prsstarter.example.com/mpc_platforms: linux/arm64,linux/ppc64le,linux/s390x` annotation, but `linux/s390x` was the only one that was not available. Code looks like this:

```
def can_pr_be_started(pr_required_platforms, mpc_available_slots)
    for platform in pr_required_platforms:
        if mpc_available_slots[platform] <= 0:
            return False
    return True
```

> It seems like we would need to add some sort of queuing mechanism for the pending PipelineRuns. How can we ensure fairness with all of the pipelines which may be waiting for different metrics/resources to indicate capacity?

You are right, we need to be reconciling (i.e. possibly starting) pipelineruns in order how they were created.

We can also introduce come sort of priority annotation that would also allow the pipelineruns to mess with the order (e.g. you are building security update and you do not want to wait for all ongoing rebuild of all RHEL 9.8 packages build).

> How can we ensure that a pipeline is not unnecessarily blocked by an unneeded resource just because it is waiting behind some other pipeline in a queue?

I was thinking about it in a way that each pipeline gated only by gating plugins it configures in it's annotations, so if I want to be gated by MPC, I use `prsstarter.example.com/mpc_platforms`, if I want to be gated by something else, I use `prsstarter.example.com/some_other_plugin`

And we can also have some global, implicit gateways like that etcd protection one that would be enabled for all pending pipelines by default.

Of course this all would only apply to pipeline runs created in this Pending state, so normal pipelines would just go ahead with kinda priority, as they would not be restricted by this gating.

Reviewer 3: >  It seems like we would need to add some sort of queuing mechanism for the pending PipelineRuns. How can we ensure fairness with all of the pipelines which may be waiting for different metrics/resources to indicate capacity?

We could do a simple "FIFO" based on the creation dates.

Reviewer: This is certainly an option, but it would result in a single-arch build staying in pending until multi-arch builds at the queue HEAD have sufficient capacity scheduled? 

Or would there be separate queues for drastically different types of requests?

> Currently this is POC code only and besides lots of other things, we know about these issues that would need to happen:

Reviewer: You refer to code but don't reference it. Code isn't necessary for an ADR as we are instead trying to understand the motivation behind a problem and decide on the defining properties of functionality.

Author: This is the code I have so far: https://gitlab.cee.redhat.com/jhutar/mystoneinst/-/blob/main/prs-controller.py?ref_type=heads

I just wanted to note that there are some blockers and I can also mention some problems I hit when creating and testing the code where I only have workarounds for them, not solutions.

Or I can just remove this section?
